### DISCUSSION TITLE
How to upload files?

### DISCUSSION THREAD

**Discussion by zwl1619** [blitz-js/blitz#1454] [Q&A] [answered]
How to upload files? Such as avatars or other images or files.

**Comment by flybayer** [comment, marked as answer]
@zwl1619 where do you want to upload the files to?

**Reply by zwl1619** [reply]
@flybayer  For example, I write a blog system with blitz, I deploy the blog system on my own Linux server, I need to upload pictures to the server when writing an article.

**Reply by flybayer** [reply]
@zwl1619 ok then you'll want to use an [API Route](https://blitzjs.com/docs/api-routes) that you can submit your file to and save it directly, like you would do with any Node.js server.

**Reply by aniss** [reply]
Hi @flybayer, is there a reason for using API Route instead of a mutation resolver? I think there's a confusion around what goes into mutations. Looking to the documentation it says you can use any node code.

**Reply by flybayer** [reply]
@aniss in mutations you can use any node code, but you don't have direct access to the node.js request/response objects. If you need direct access to those, you have to use API routes (or add HTTP middleware which is discouraged, but file upload could be a good use case for that).

By the way, we also have [this open issue](https://github.com/blitz-js/blitz/issues/843) for adding file upload support directly to mutations

**Reply by aniss** [reply]
that make sense now.

**Reply by opensourcekam** [reply]
@flybayer it would be awesome if you could access the mutation / query resolvers similar to how you can with Apollo/ Yoga

```
t.field('uploadFile', {
			type: 'File',
			args: {
				public_id: stringArg(),
				secret: stringArg(),
				name: stringArg(),
				url: stringArg(),
				contentType: stringArg(),
				size: intArg(),
				width: intArg(),
				height: intArg()
			},
			resolve: async (_, { public_id, secret, name, url, size, width, height, contentType }, ctx) => {
				const userId = getUserId(ctx);

				return await ctx.prisma.createFile({
					public_id,
					secret,
					name,
					url,
					size,
					width,
					height,
					contentType,
					author: {
						connect: { id: userId }
					}
				});
			}
		});
```

Similar to such

**Reply by flybayer** [reply]
@opensourcekam access the resolvers from where? Your code example will directly port to a a Blitz mutation.

**Reply by krunaldodiya** [reply]
How about writing a wrapper around https://github.com/Slynova-Org/flydrive, laravel also use flysystem api for managing files for different storage adapters like local, aws s3, digital ocean space, google cloud storage etc
